internal/app/apid/backend.go

```diff
@@ -0,0 +1,116 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/talos-systems/grpc-proxy/proxy"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/metadata"
+
+	"github.com/talos-systems/talos/api/common"
+	"github.com/talos-systems/talos/pkg/constants"
+)
+
+type apidBackend struct {
+	Target string
+	Creds  credentials.TransportCredentials
+
+	mu   sync.Mutex
+	conn *grpc.ClientConn
+}
+
+func (ab *apidBackend) String() string {
+	return ab.Target
+}
+
+// GetConnection returns a grpc connection to the backend.
+func (ab *apidBackend) GetConnection(ctx context.Context) (context.Context, *grpc.ClientConn, error) {
+	origMd, ok := metadata.FromIncomingContext(ctx)
+
+	md := origMd.Copy()
+	delete(md, "targets")
+
+	if ok {
+		md.Set("proxyfrom", origMd[":authority"]...)
+	} else {
+		md.Set("proxyfrom", "unknown")
+	}
+
+	outCtx := metadata.NewOutgoingContext(ctx, md)
+
+	ab.mu.Lock()
+	defer ab.mu.Unlock()
+
+	if ab.conn != nil {
+		return outCtx, ab.conn, nil
+	}
+
+	var err error
+	ab.conn, err = grpc.DialContext(
+		ctx,
+		fmt.Sprintf("%s:%d", ab.Target, constants.OsdPort),
+		grpc.WithTransportCredentials(ab.Creds),
+		grpc.WithCodec(proxy.Codec()), //nolint: staticcheck
+	)
+
+	log.Printf("got conn, %v %v", ab.conn, err)
+
+	return outCtx, ab.conn, err
+}
+
+// AppendInfo is called to enhance response from the backend with additional data.
```

Comment:
need better docs on what is going here

Reply:
i _think_ i get what's going on here but it definitely feels like its in the realm of black wizardry,

Reply:
New version has docstring & tests